src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -230,41 +232,98 @@ private static (SyntaxTriviaList sameLine, SyntaxTriviaList endOfLine) SplitTrai
+                    {
+                        firstVariableUsage[localSymbol] = statementIndex;
+                    }
+
+                    // Add the variables to the list of locals.
+                    localVariables.AddRange(localSymbols);
+                }
+            }
+
+            // If the trigger variable has no other usage, then return null.
+            if (!lastVariableUsage.ContainsKey(localVariable))
+            {
+                return null;
+            }
+
+            var lastUsageIndex = lastVariableUsage[localVariable];
+
+            // Skip the trigger variable and loop through local variables
+            // in the order that they were declared.
+            foreach (var localSymbol in localVariables.Skip(1))
+            {
+                var declarationIndex = firstVariableUsage[localSymbol];
+                if (declarationIndex > lastUsageIndex)
                 {
-                    break; // Ignore the declaration and usages prior to the declaration
+                    // If the variable was declared after the last statement to include in
+                    // the using statement, we have gone far enough and other variables should
+                    // also be outside the using statement.
+                    break;
                 }
 
-                if (ContainsReference(node, localVariable, semanticModel, syntaxFactsService, cancellationToken))
+                // If the variable has no other usages, then we can move to the next variable.
+                if (!lastVariableUsage.TryGetValue(localSymbol, out var lastUsage))
                 {
-                    return node;
+                    continue;
                 }
+
+                lastUsageIndex = Math.Max(lastUsageIndex, lastUsage);
             }
 
-            return null;
+            return statementsAfterDeclaration[lastUsageIndex];
         }
 
         private static bool ContainsReference(
             SyntaxNode node,
-            ILocalSymbol localVariable,
+            List<ILocalSymbol> localVariables,
             SemanticModel semanticModel,
             ISyntaxFactsService syntaxFactsService,
-            CancellationToken cancellationToken)
+            CancellationToken cancellationToken,
+            out ILocalSymbol referencedVariable)
         {
             if (syntaxFactsService.IsIdentifierName(node))
             {
                 var identifierName = syntaxFactsService.GetIdentifierOfSimpleName(node).ValueText;
 
-                return syntaxFactsService.StringComparer.Equals(localVariable.Name, identifierName) &&
-                    localVariable.Equals(semanticModel.GetSymbolInfo(node).Symbol);
+                referencedVariable = localVariables.FirstOrDefault(localVariable
+                    => syntaxFactsService.StringComparer.Equals(localVariable.Name, identifierName) &&
+                        localVariable.Equals(semanticModel.GetSymbolInfo(node).Symbol));
```

Comment:
firstOrDefault seems wrong.  you could have a statement that has *many* references to some o fthe locals in your localVariables list.  But you'll only update the 'lastUsage' for the first one you return here right?

Instead of ContainsReference, this should likely be something like: GetReferencesLocalVariables, right?